fdbserver/MetricLogger.actor.cpp

```diff
@@ -364,41 +375,275 @@ ACTOR Future<Void> updateMetricRegistration(Database cx, MetricsConfig* config,
+// 				break;
+// 		wait(delay(1.0));
+// 	}
+
+// 	state MetricsConfig config(prefix);
+
+// 	try {
+// 		Database cx = wait(fcx);
+// 		Future<Void> conf = metricRuleUpdater(cx, &config, metrics);
+// 		Future<Void> dump = dumpMetrics(cx, &config, metrics);
+// 		Future<Void> reg = updateMetricRegistration(cx, &config, metrics);
+
+// 		wait(conf || dump || reg);
+// 	} catch (Error& e) {
+// 		if (e.code() != error_code_actor_cancelled) {
+// 			// Disable all metrics
+// 			for (auto& it : metrics->metricMap)
+// 				it.value->setConfig(false);
+// 		}
+
+// 		TraceEvent(SevWarnAlways, "TDMetricsStopped").error(e);
+// 		throw e;
+// 	}
+// 	return Void();
+// }
+
+ACTOR Future<Void> startMetricsSimulationServer(MetricsDataModel model) {
+	if (model == MetricsDataModel::NONE) {
+		return Void{};
 	}
+	state uint32_t port = (model == STATSD) ? FLOW_KNOBS->STATSD_UDP_EMISSION_PORT : FLOW_KNOBS->OTEL_UDP_EMISSION_PORT;
+	TraceEvent(SevInfo, "MetricsUDPServerStarted").detail("Address", "127.0.0.1").detail("Port", port);
+	state NetworkAddress localAddress = NetworkAddress::parse("127.0.0.1:" + std::to_string(port));
+	state Reference<IUDPSocket> serverSocket = wait(INetworkConnections::net()->createUDPSocket(localAddress));
+	serverSocket->bind(localAddress);
+	state Standalone<StringRef> packetString = makeString(IUDPSocket::MAX_PACKET_SIZE);
+	state uint8_t* packet = mutateString(packetString);
 
-	// Wait until the collection has been created and initialized.
-	state TDMetricCollection* metrics = nullptr;
 	loop {
-		metrics = TDMetricCollection::getTDMetrics();
-		if (metrics != nullptr)
-			if (metrics->init())
-				break;
-		wait(delay(1.0));
+		int size = wait(serverSocket->receive(packet, packet + IUDPSocket::MAX_PACKET_SIZE));
+		auto message = packetString.substr(0, size);
+
+		// Let's just focus on statsd for now. For statsd, the message is expected to be seperated by newlines. We need
+		// to break each statsd metric and verify them individually.
+		if (model == MetricsDataModel::STATSD) {
+			std::string statsd_message = message.toString();
+			auto metrics = splitString(statsd_message, "\n");
+			for (const auto& metric : metrics) {
+				ASSERT(verifyStatsdMessage(metric));
+			}
+		}
 	}
+}
 
-	state MetricsConfig config(prefix);
+ACTOR Future<Void> runMetrics() {
+	state MetricCollection* metrics = nullptr;
+	state MetricBatch batch;
+	MetricsDataModel model = knobToMetricModel(FLOW_KNOBS->METRICS_DATA_MODEL);
+	if (model == MetricsDataModel::NONE) {
+		return Void{};
+	}
+	state UDPMetricClient metricClient;
+	state Future<Void> metrics_actor;
+	if (g_network->isSimulated()) {
+		metrics_actor = startMetricsSimulationServer(model);
+	}
+	loop {
+		metrics = MetricCollection::getMetricCollection();
+		if (metrics != nullptr) {
+			for (const auto& p : metrics->map) {
+				p.second->flush(batch);
+			}
+			metricClient.send(batch);
+		}
+		batch.clear();
+		wait(delay(FLOW_KNOBS->METRICS_EMISSION_INTERVAL));
```

Comment:
I think we need to rethink this style of metrics emission. Ignoring statsd for now, having an emission loop that runs separately from how metrics currently write themselves to `TraceEvent`s causes complication and other problems such as needing to make a copy.

What about adding a new `MetricDataModel` type of `TraceEvent`? Then in the `flush` function of each metric implementation, the metric could be written to trace events as well as to OTEL. This would require metrics to be able to have multiple `MetricDataModel`s, and be able to flush multiple at once. Then, we wouldn't need separate actors to do the log based flushing.

This loop would likely have to change. But I think a design where each metric defines its own flush interval, and then the main loop flushes it on that interval to as many "outputs" as required, makes the most sense. Then, each metric can decide if it wants to flush itself to logs, OTEL, both, none, etc...